File: 第五阶段/5.2-多线程与GIL/ThreadDemo.cs
// =============================================================================
// C# 多线程示例 — C# 老兵的 Python 修炼手册 5.2
// 对应文章：5.2 多线程与GIL
// 关键概念：Task, Thread, ThreadPool, async, Parallel,
//           无 GIL 的真正并行
// =============================================================================
// 【C# vs Python 对比 — 最大的差异之一】
//
// C# 没有 GIL（全局解释器锁），线程可以真正并行执行 CPU 密集型任务
// Python 有 GIL，同一时刻只有一个线程执行 Python 字节码
//
// 这意味着 C# 在多线程方面天然强于 Python：
// - C# Task.Run 即可利用多核
// - Python 需要 multiprocessing 才能绕过 GIL
// =============================================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace ThreadDemo
{
    public static class ThreadExamples
    {
        private static int _counter = 0;
        private static readonly object _lockObj = new object();

        // =========================================================================
        // 1. C# 没有 GIL —— 线程可以真正并行
        // =========================================================================
        // 【Python 对比】
        // Python 的 threading 受 GIL 限制，CPU 密集型任务无法加速
        // C# 的线程由 CLR 管理，每个线程可以独立执行 JIT 编译的机器码

        /// <summary>
        /// 无锁原子操作
        /// 【Python 对比】需要 threading.Lock() 手动加锁
        /// C# 的 Interlocked 类提供原子操作，不需要锁
        /// </summary>
        public static void IncrementWithInterlocked()
        {
            for (int i = 0; i < 1_000_000; i++)
            {
                Interlocked.Increment(ref _counter);
            }
        }

        /// <summary>
        /// 使用 lock 语句
        /// 【Python 对比】with threading.Lock(): counter += 1
        /// </summary>
        public static void IncrementWithLock()
        {
            for (int i = 0; i < 1_000_000; i++)
            {
                lock (_lockObj)
                {
                    _counter++;
                }
            }
        }

        /// <summary>
        /// 演示线程安全的并发递增
        /// </summary>
        public static void DemoThreadingBasics()
        {
            Console.WriteLine("\n=== 1. 线程安全的并发操作 ===");

            _counter = 0;
            var t1 = new Thread(IncrementWithInterlocked);
            var t2 = new Thread(IncrementWithInterlocked);

            var sw = Stopwatch.StartNew();
            t1.Start();
            t2.Start();
            t1.Join();
            t2.Join();
            sw.Stop();

            Console.WriteLine($"  两个线程各递增 1,000,000 次");
            Console.WriteLine($"  结果: {Interlocked(ref _counter):N0} (期望: 2,000,000)");
            Console.WriteLine($"  耗时: {sw.ElapsedMilliseconds}ms");
        }

        private static int Interlocked(ref int target) => target;

        // =========================================================================
        // 2. C# 没有 GIL —— CPU 密集型任务可以真正加速
        // =========================================================================
        // 【Python 对比】
        // Python 的 threading 对 CPU 密集型无效（GIL 限制）
        // C# 的 Task.Run / Parallel.For 可以真正利用多核

        /// <summary>
        /// CPU 密集型计算
        /// </summary>
        public static double CpuBoundWork(int n)
        {
            double total = 0;
            for (int i = 0; i < n; i++)
            {
                total += Math.Sqrt(i);
            }
            return total;
        }

        /// <summary>
        /// 演示 CPU 密集型任务的真正并行
        /// 【Python 对比】需要用 multiprocessing 才能达到类似效果
        /// </summary>
        public static void DemoNoGilParallel()
        {
            Console.WriteLine("\n=== 2. 没有 GIL — CPU 密集型真正并行 ===");

            int n = 2_000_000;

            // 单线程串行
            var sw = Stopwatch.StartNew();
            double r1 = CpuBoundWork(n);
            double r2 = CpuBoundWork(n);
            long singleTime = sw.ElapsedMilliseconds;
            Console.WriteLine($"  单线程串行: {singleTime}ms (结果: {r1 + r2:N0})");

            // 多线程并行（C# 真正利用多核！）
            sw.Restart();
            double[] results = new double[2];
            Parallel.For(0, 2, i =>
            {
                results[i] = CpuBoundWork(n);
            });
            long parallelTime = sw.ElapsedMilliseconds;
            Console.WriteLine($"  Parallel.For: {parallelTime}ms (结果: {results.Sum():N0})");
            Console.WriteLine($"  加速比: {(double)singleTime / parallelTime:F2}x");
            Console.WriteLine($"  【Python 对比】Python threading 无法做到这一点（GIL 限制）");
        }

        // =========================================================================
        // 3. Task —— C# 最常用的并发方式
        // =========================================================================
        // 【Python 对比】
        // C#: Task.Run(() => DoWork())
        // Python: threading.Thread(target=func).start() 或 asyncio.create_task(coro)

        /// <summary>
        /// 演示 Task 的使用
        /// </summary>
        public static async Task DemoTaskAsync()
        {
            Console.WriteLine("\n=== 3. Task 并发 ===");

            // 创建多个并行任务
            var tasks = Enumerable.Range(1, 5)
                .Select(i => Task.Run(() =>
                {
                    Thread.Sleep(200);  // 模拟工作
                    return $"任务{i}完成 (线程ID: {Thread.CurrentThread.ManagedThreadId})";
                }))
                .ToList();

            string[] results = await Task.WhenAll(tasks);
            foreach (string result in results)
            {
                Console.WriteLine($"  {result}");
            }

            Console.WriteLine($"  主线程 ID: {Thread.CurrentThread.ManagedThreadId}");
            Console.WriteLine($"  【Python 对比】asyncio.gather() 实现类似功能");
        }

        // =========================================================================
        // 4. ThreadPool —— 线程池管理
        // =========================================================================
        // 【Python 对比】concurrent.futures.ThreadPoolExecutor
        // C# 的 ThreadPool 由 CLR 自动管理

        /// <summary>
        /// 演示 ThreadPool 和 Parallel
        /// </summary>
        public static void DemoThreadPool()
        {
            Console.WriteLine("\n=== 4. ThreadPool 与 Parallel ===");

            // ThreadPool.QueueUserWorkItem —— 底层的线程池任务
            // 【Python 对比】executor.submit(func, args)
            var resetEvent = new ManualResetEventSlim(false);

            ThreadPool.QueueUserWorkItem(_ =>
            {
                Console.WriteLine($"  ThreadPool 工作项 (线程ID: {Thread.CurrentThread.ManagedThreadId})");
                resetEvent.Set();
            });

            resetEvent.Wait();

            // Parallel.For —— 数据并行
            // 【Python 对比】multiprocessing.Pool.map() 或 concurrent.futures
            Console.WriteLine("  Parallel.For 结果:");
            long sum = 0;
            Parallel.For(0, 100, i =>
            {
                Interlocked.Add(ref sum, i);
            });
            Console.WriteLine($"    0到99的和 = {sum}");
        }

        // =========================================================================
        // 5. 异步 I/O —— 对比 Python asyncio
        // =========================================================================
        // 【Python 对比】
        // Python: async def fetch(url): ... await asyncio.sleep(1)
        // C#:     async Task<string> FetchAsync(url) { await Task.Delay(1000); }
        //
        // 两者原理相同：不阻塞线程，让出 CPU 给其他任务

        /// <summary>
        /// 模拟异步 I/O 操作
        /// </summary>
        public static async Task<string> FetchDataAsync(string url, int delayMs)
        {
            Console.WriteLine($"  [开始] {url}");
            await Task.Delay(delayMs);  // 不阻塞线程
            Console.WriteLine($"  [完成] {url}");
            return $"来自 {url} 的数据";
        }

        /// <summary>
        /// 演示异步 I/O 并发
        /// </summary>
        public static async Task DemoAsyncIoAsync()
        {
            Console.WriteLine("\n=== 5. 异步 I/O（对比 Python asyncio）===");

            var sw = Stopwatch.StartNew();

            // 并发执行多个异步请求
            // 【Python 对比】await asyncio.gather(fetch(u) for u in urls)
            var tasks = new[]
            {
                FetchDataAsync("api/users", 1000),
                FetchDataAsync("api/orders", 1500),
                FetchDataAsync("api/products", 800)
            };

            string[] results = await Task.WhenAll(tasks);

            sw.Stop();
            Console.WriteLine($"  获取 {results.Length} 个结果, 耗时: {sw.ElapsedMilliseconds}ms");
            Console.WriteLine($"  【Python 对比】asyncio.gather() 实现相同效果");
        }

        // =========================================================================
        // 6. CancellationToken —— 对比 Python 的取消机制
        // =========================================================================
        // 【Python 对比】
        // Python: asyncio 通过 CancelledError 取消协程
        // C#: CancellationToken 协作式取消，更优雅

        /// <summary>
        /// 可取消的异步任务
        /// </summary>
        public static async Task CancellableWorkAsync(CancellationToken token)
        {
            for (int i = 0; i < 10; i++)
            {
                token.ThrowIfCancellationRequested();
                Console.WriteLine($"  工作步骤 {i + 1}/10");
                await Task.Delay(200, token);
            }
        }

        /// <summary>
        /// 演示取消操作
        /// </summary>
        public static async Task DemoCancellationAsync()
        {
            Console.WriteLine("\n=== 6. CancellationToken（协作式取消）===");

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));

            try
            {
                await CancellableWorkAsync(cts.Token);
            }
            catch (OperationCanceledException)
            {
                Console.WriteLine("  [已取消] 任务在1秒后被取消");
            }

            Console.WriteLine("  【Python 对比】asyncio 的 CancelledError 机制类似");
        }
    }

    // =========================================================================
    // 主程序入口
    // =========================================================================
    class Program
    {
        /// <summary>
        /// 程序入口
        /// 【Python 对比】if __name__ == "__main__": main()
        /// </summary>
        static async Task Main(string[] args)
        {
            Console.WriteLine("============================================================");
            Console.WriteLine("  C# 老兵的 Python 修炼手册 — 5.2 多线程与 GIL");
            Console.WriteLine("============================================================");
            Console.WriteLine($"  C# CLR 没有 GIL，线程可以真正并行");
            Console.WriteLine($"  Python CPython 有 GIL，CPU 密集型需用 multiprocessing");

            ThreadExamples.DemoThreadingBasics();
            ThreadExamples.DemoNoGilParallel();
            await ThreadExamples.DemoTaskAsync();
            ThreadExamples.DemoThreadPool();
            await ThreadExamples.DemoAsyncIoAsync();
            await ThreadExamples.DemoCancellationAsync();

            Console.WriteLine("\n============================================================");
            Console.WriteLine("  核心结论：");
            Console.WriteLine("  1. C# 没有 GIL，Task.Run / Parallel.For 天然支持并行");
            Console.WriteLine("  2. Python 有 GIL，CPU 密集型必须用 multiprocessing");
            Console.WriteLine("  3. I/O 密集型两者都可用异步模型处理");
            Console.WriteLine("  4. C# 的 CancellationToken 比 Python 的取消机制更优雅");
            Console.WriteLine("============================================================");
        }
    }
}

File: 第五阶段/5.2-多线程与GIL/thread_demo.py
# =============================================================================
# Python 多线程与 GIL 示例 — C# 老兵的 Python 修炼手册 5.2
# 对应文章：5.2 多线程与GIL
# 关键概念：threading, GIL限制, multiprocessing, concurrent.futures,
#           asyncio vs threading vs multiprocessing
# =============================================================================
# 【C# vs Python 对比 — 最大的差异之一】
#
# C# 没有 GIL（全局解释器锁），线程可以真正并行执行 CPU 密集型任务
# Python 有 GIL，同一时刻只有一个线程执行 Python 字节码
#
# C# 多线程: Task.Run(() => { ... }) -> 线程池中的真实并行
# Python 多线程: threading.Thread(target=fn) -> 受 GIL 限制的并发
#
# 结论: Python CPU 密集型任务用 multiprocessing，I/O 密集型用 asyncio
# =============================================================================

import threading
import multiprocessing
import concurrent.futures
import time
import sys
from typing import List


# =============================================================================
# 1. GIL 是什么？—— 全局解释器锁
# =============================================================================
# GIL (Global Interpreter Lock) 是 CPython 解释器中的一个互斥锁，
# 保证同一时刻只有一个线程执行 Python 字节码。
#
# 【C# 对比】C# CLR 没有 GIL，每个线程可以独立执行 JIT 编译的机器码
# 这意味着 C# 的多线程可以真正利用多核 CPU 进行并行计算
#
# 为什么 Python 需要 GIL？
# - CPython 使用引用计数进行内存管理
# - GIL 简化了内存管理，避免了竞态条件
# - 但也意味着 CPU 密集型任务无法通过多线程加速


def demo_what_is_gil():
    """解释 GIL 的概念"""
    print("\n=== 1. 什么是 GIL ===")
    print(f"  Python 实现: {sys.implementation.name}")
    print(f"  Python 版本: {sys.version}")
    print(f"  GIL 存在: 是 (仅 CPython，PyPy 有 GIL-free 模式)")
    print(f"  【C# 对比】C# CLR 没有 GIL，线程可以真正并行")
    print(f"  Python 3.12+ 有 PEP 684 子解释器独立 GIL (实验性)")
    print(f"  Python 3.13+ 支持 free-threaded 模式 (实验性)")


# =============================================================================
# 2. threading 模块 —— I/O 密集型任务
# =============================================================================
# 【C# 对比】
# C#: Task.Run(() => DoWork()) 或 new Thread(() => DoWork()).Start()
# Python: threading.Thread(target=func).start()
#
# 两者都能用于 I/O 密集型任务（网络请求、文件读写、数据库查询）
# 因为 I/O 等待期间线程会释放 GIL（Python）或让出 CPU（C#）

shared_counter = 0
counter_lock = threading.Lock()


def increment_shared(n: int) -> None:
    """线程安全的递增操作"""
    global shared_counter
    for _ in range(n):
        # 不加锁会出问题（GIL 不能保证原子操作）
        with counter_lock:
            shared_counter += 1


def demo_threading_basics():
    """演示 threading 基础"""
    print("\n=== 2. threading 模块（I/O 密集型） ===")

    # 【C# 对比】
    # var t1 = new Thread(() => IncrementShared(1000000));
    # var t2 = new Thread(() => IncrementShared(1000000));
    # t1.Start(); t2.Start();
    # t1.Join(); t2.Join();

    n = 1_000_000
    t1 = threading.Thread(target=increment_shared, args=(n,))
    t2 = threading.Thread(target=increment_shared, args=(n,))

    start = time.perf_counter()
    t1.start()
    t2.start()
    t1.join()
    t2.join()
    elapsed = time.perf_counter() - start

    print(f"  两个线程各递增 {n:,} 次")
    print(f"  结果: {shared_counter:,} (期望: {2 * n:,})")
    print(f"  耗时: {elapsed:.3f}s")


# =============================================================================
# 3. GIL 的影响 —— CPU 密集型任务不会加速
# =============================================================================
# 这是 C# 开发者最容易踩的坑！
# 在 C# 中，Parallel.For 会让 CPU 密集型任务加速
# 在 Python 中，threading 反而可能更慢（线程切换开销 + GIL）

def cpu_bound_work(n: int) -> float:
    """CPU 密集型计算"""
    total = 0.0
    for i in range(n):
        total += i ** 0.5
    return total


def demo_gil_limitation():
    """演示 GIL 对 CPU 密集型任务的限制"""
    print("\n=== 3. GIL 限制 — CPU 密集型任务 ===")

    n = 2_000_000

    # 单线程
    start = time.perf_counter()
    r1 = cpu_bound_work(n)
    r2 = cpu_bound_work(n)
    single_time = time.perf_counter() - start
    print(f"  单线程: {single_time:.3f}s (结果: {r1 + r2:.0f})")

    # 多线程（受 GIL 限制，可能不会更快甚至更慢）
    start = time.perf_counter()
    with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
        futures = [executor.submit(cpu_bound_work, n) for _ in range(2)]
        results = [f.result() for f in futures]
    threaded_time = time.perf_counter() - start
    print(f"  多线程: {threaded_time:.3f}s (结果: {sum(results):.0f})")

    ratio = single_time / threaded_time if threaded_time > 0 else 0
    if ratio < 1.0:
        print(f"  [GIL 限制] 多线程反而慢了 {1/ratio:.2f}x — GIL 导致!")
    else:
        print(f"  多线程略快 {ratio:.2f}x (I/O 开销可能被 GIL 掩盖)")

    print("  【C# 对比】C# Parallel.For 可以真正加速，因为没有 GIL")


# =============================================================================
# 4. multiprocessing —— 绕过 GIL，真正并行
# =============================================================================
# 【C# 对比】
# C#: Parallel.For(0, 4, i => { results[i] = CpuBound(i); });
#      或 Task.Run(() => CpuBound(i))
#
# Python multiprocessing: 每个子进程有独立的 Python 解释器和 GIL
#                         真正的并行，但进程间通信有开销

def cpu_bound_process(n: int) -> float:
    """在子进程中执行的 CPU 密集型计算"""
    return sum(i ** 0.5 for i in range(n))


def demo_multiprocessing():
    """演示 multiprocessing 绕过 GIL"""
    print("\n=== 4. multiprocessing（绕过 GIL，真正并行） ===")

    n = 2_000_000
    num_workers = 4

    # 单进程
    start = time.perf_counter()
    single_result = sum(cpu_bound_process(n) for _ in range(num_workers))
    single_time = time.perf_counter() - start
    print(f"  单进程 ({num_workers}次串行): {single_time:.3f}s")

    # 多进程
    start = time.perf_counter()
    with multiprocessing.Pool(num_workers) as pool:
        results = pool.map(cpu_bound_process, [n] * num_workers)
    multi_time = time.perf_counter() - start
    print(f"  多进程 ({num_workers}核并行): {multi_time:.3f}s")
    print(f"  加速比: {single_time / multi_time:.2f}x")
    print(f"  【C# 对比】C# 的 Parallel.For 天然支持并行，不需要额外模块")


# =============================================================================
# 5. concurrent.futures —— 统一的并发接口
# =============================================================================
# 【C# 对比】
# C# 的 Task + Task.WhenAll 是统一的异步模型
# Python 的 concurrent.futures 提供 ThreadPoolExecutor 和 ProcessPoolExecutor
#
# ThreadPoolExecutor  → 用于 I/O 密集型（类似 C# Task.Run）
# ProcessPoolExecutor → 用于 CPU 密集型（类似 C# Parallel.For）

def demo_concurrent_futures():
    """演示 concurrent.futures 模块"""
    print("\n=== 5. concurrent.futures（统一并发接口） ===")

    # ThreadPoolExecutor —— I/O 密集型
    # 【C# 对比】Task.WhenAll
    def io_task(task_id: int) -> str:
        time.sleep(0.3)  # 模拟 I/O
        return f"任务{task_id}完成"

    start = time.perf_counter()
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        # 提交多个任务
        futures = {executor.submit(io_task, i): i for i in range(5)}
        for future in concurrent.futures.as_completed(futures):
            task_id = futures[future]
            result = future.result()
            print(f"  {result}")
    io_time = time.perf_counter() - start
    print(f"  I/O 任务总耗时: {io_time:.3f}s (并发执行)")

    # ProcessPoolExecutor —— CPU 密集型
    # 【C# 对比】Parallel.For
    with concurrent.futures.ProcessPoolExecutor(max_workers=2) as executor:
        results = list(executor.map(cpu_bound_process, [500_000] * 2))
    print(f"  CPU 任务结果: [{len(results)} 个子进程完成]")


# =============================================================================
# 6. asyncio vs threading vs multiprocessing —— 选择指南
# =============================================================================

def demo_comparison():
    """对比三种并发模型"""
    print("\n=== 6. asyncio vs threading vs multiprocessing 选择指南 ===")
    print("  ┌─────────────────┬──────────────┬──────────────┬──────────────┐")
    print("  │ 特性            │ asyncio      │ threading    │ multiprocessing │")
    print("  ├─────────────────┼──────────────┼──────────────┼──────────────┤")
    print("  │ GIL 影响        │ 不受影响     │ 受限         │ 不受限       │")
    print("  │ 适用场景        │ I/O 密集     │ I/O 密集     │ CPU 密集     │")
    print("  │ 内存共享        │ 是           │ 是(需加锁)   │ 否(需IPC)    │")
    print("  │ 创建开销        │ 最小         │ 中等         │ 最大         │")
    print("  │ 切换开销        │ 最小         │ 中等         │ 最大         │")
    print("  │ C# 对应         │ async/await  │ Task.Run     │ Parallel.For │")
    print("  └─────────────────┴──────────────┴──────────────┴──────────────┘")


# =============================================================================
# 7. 实际案例：I/O 密集型并发下载
# =============================================================================

def demo_io_concurrent():
    """实际案例：模拟并发下载"""
    print("\n=== 7. 实际案例：I/O 密集型并发 ===")

    def download_page(url: str) -> dict:
        """模拟下载网页"""
        time.sleep(0.2)  # 模拟网络延迟
        return {"url": url, "size": len(url) * 100}

    urls = [f"https://example.com/page{i}" for i in range(10)]

    # 串行
    start = time.perf_counter()
    serial_results = [download_page(url) for url in urls]
    serial_time = time.perf_counter() - start
    print(f"  串行: {serial_time:.3f}s")

    # 并发（threading + futures）
    start = time.perf_counter()
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        parallel_results = list(executor.map(download_page, urls))
    parallel_time = time.perf_counter() - start
    print(f"  并发: {parallel_time:.3f}s (5个线程)")
    print(f"  加速比: {serial_time / parallel_time:.1f}x")


# =============================================================================
# 主程序入口
# =============================================================================

def main():
    """运行所有演示
    
    【C# 对比】
    C#: static void Main() { DemoAll(); }
    Python: main() + if __name__ == "__main__"
    """
    print("=" * 60)
    print("  C# 老兵的 Python 修炼手册 — 5.2 多线程与 GIL")
    print("=" * 60)

    demo_what_is_gil()
    demo_threading_basics()
    demo_gil_limitation()
    demo_multiprocessing()
    demo_concurrent_futures()
    demo_comparison()
    demo_io_concurrent()

    print("\n" + "=" * 60)
    print("  核心结论：")
    print("  1. Python 有 GIL，多线程不能加速 CPU 密集型任务")
    print("  2. CPU 密集型用 multiprocessing，I/O 密集型用 asyncio")
    print("  3. C# 没有 GIL，Task.Run + Parallel.For 可以真正并行")
    print("  4. asyncio 是 Python 处理高并发 I/O 的最佳选择")
    print("=" * 60)


if __name__ == "__main__":
    main()
